### DISCUSSION TITLE
Living Documentation

### DISCUSSION THREAD

**Discussion by Code-Grump** [reqnroll/Reqnroll#196] [Ideas]
Having seen a few mentions of the SpecFlow Living Docs plugin, it seems like there's a demand for an equivalent feature in Reqnroll.

What I'd like to understand is:
- Does it matter when this output is produced?
- Should it be the output of a build, or a test-run?
- Would running a stand-alone tool to extract this information be preferrable to something built into Visual Studio?
- What do we think should be the output?

I can think of many ways to approach this, so knowing the constraints is probably the easiest way to find a path forward.

**Comment by gasparnagy** [comment]
Thx for taking care of this.

I don't think there are single answers for these questions (there are so many situation and needs), but let me share my thoughts based the usages of SpecFlow+ LivingDoc I have seen. 

> Does it matter when this output is produced?

The most common usage is to produce the report at the same time of the test execution or directly after in CI (or locally). SpecFlow+ LivingDoc CLI enforced a separate step (generated the data during test execution and run the HTML gen after), but I know that many teams tried to merge this by calling the HTML gen from an AfterTestRun hook.

> Should it be the output of a build, or a test-run?

I think the most common usage is to use it with test-run, but I have seen usages where they were only interested in the nice rendering of the feature file and not interested in the test result. So both, but more commonly related to test-run.

> Would running a stand-alone tool to extract this information be preferrable to something built into Visual Studio?

Yes, definitely, as it is going to be used mostly as part of the CI. I would even say a "stand-alone command line tool" (no UI).

> What do we think should be the output?

I think there are diverse answers for this. The most simple and common usage would require a single-file active HTML (maybe multi-file HTML is also acceptable, but more problematic to share). But there are teams who use this as part of their QA process, where they might need to produce other output (PDF, Word, ???). But I think this latter is more the special case.

When I say active HTML, I mainly mean that something where you can browse the folders and feature files, search, etc. within the offline HTML. 

Besides these questions, there are plenty of "nice-to-haves" that I have heard of. These are not necessarily part of the primary goal (or a goal at all), but good to think about them when we set the design. (SpecFlow+ LivinDoc supported some of these.)

* Customizable HTML output (templating?)
* Have summary controls (charts) in the result
* Non-HTML output (PDF, Word, JSON)
* Seamless integration of the result document to common ALM systems, like Azure DevOps, Jira, GitHub
* Manage result history

**Reply by tula-engineered-quality-hub** [reply]
Hi @gasparnagy! How is the live documentation part going? Publishing the HTML extension on Azure is a prerequisite for us to finally migrate from Specflow to Reqnroll and take advantage of all the improvements made. I must admit, I'm very excited about this. 

I would like to understand the status of this request, if there is anything in progress, any planning, and if https://github.com/praticandobdd and I can actively contribute (if so, what contribution would add value), especially in terms of development and testing. We have a rich environment for testing this entire process, as well as experience to share with this community.

**Reply by gasparnagy** [reply]
@clrudolphi is currently working on the cucumber messages support that is a prerequisite of this (see https://github.com/reqnroll/Reqnroll/pull/233), but we realized that it is a bigger task than we (or at least I) expected. Once this is done, we can work on the HTML generation. 

In principle we could work on the HTML generation even before, because we can get sample Cucumber Messages reports from other Cucumber implementations. But to start with that, we would need some design concept at least. My preference would be to use React (like the old one did), so that we could reuse something from https://github.com/cucumber/react-components. But that's all I have and I'm not a React expert. If you (or anyone else) is fluent in React, you could start working on a prototype...

**Reply by blaisep** [reply]
Hi @gasparnagy ... (I just learned about this proiect from Lisa Crispin) 
If there is support for python, then you can use the tooling like Sphinx or MKdocs that currently power https://readthedocs.io 
Sphinx will output to many formats (PDF, Confluence, ePub, Man Pages, HTML, etc) and if you only need HTML then there are many more options.

The tricky thing about live docs is not so much the HTML design but rather the mapping of source code to abstractions like definitions (Glossary, API, etc) and other cross references. In this sense Sphinx, AsciiDoctor and Antora would be the tools to consider.

**Reply by gasparnagy** [reply]
@blaisep Thx for the comment. Yes, that's why we currently focus on generating cucumber messages that contain all these information. Once we have that we can find the right tooling to produce HTML. Do you know maybe any tool that can generate report from Cucumber Messages `.ndjson` file directly?

**Comment by a-korevaar** [comment]
Having a LivingDoc equivalent feature would allow us to migrate from SpecFlow to Reqnroll. I would gladly contribute to the development or testing of such a feature.

**Reply by Code-Grump** [reply]
Can you tell us what you need the output to be? Any technical constraints?

**Reply by a-korevaar** [reply]
We're currently using the SpecFlow Azure DevOps extension and the CLI to generate LivingDoc in HTML format and publish it externally. I have added some custom code to the build pipeline to include images (mainly screen mockups) as well.

For us, the integration in Azure DevOps is a nice to have, as developers are focused on the feature and step files in their local repositories and product owners and stakeholders use the published documentation.

The HTML format is ok, but it would be nice to have other options such as markdown or pdf as well.

I'm not familiar with [Cucumber reports](https://github.com/orgs/reqnroll/discussions/37) but it looks interesting. We are for the most part bound to on-premises solutions though.

**Reply by Code-Grump** [reply]
> We are for the most part bound to on-premises solutions though.

This is good to know as a constraint. I'm certain it's a common one too.

**Reply by gasparnagy** [reply]
I agree. Test results are highly confidential in many environments I worked with and storing any test result data introduces a lot of complication (GDPR, etc.).

**Comment by adrichem** [comment]
I love this is getting attention! LivingDocs reports are the only useful source of test results for us. Especially the test runner mixes output from concurrently running scenarios. 

1. We need to generate and store results in our CI/CD pipelines .My org needs flexibility in deciding where/when this work is done as we have different projects using different setups in their pipelines. Some are YAML, some are classic. Some are build- some are release. Some pipelines only run tests and have no 'real' branch associated with them. Having LivingDoc invoked by shell script ticks all our boxes w.r.t. integration into our devops. 
3. Having HTML output is critical for us as we include the HTML into viewers configured in DevOps on the pipelines..
4. Its very nice LivingDocs contains the markdown from the scenarios. This helps communicate why tests do what they do.
5. When running locally we sometimes manually create a report (both for test results but also to show the content of scenario to someone in a nice format). So ability to execute multi platform  is needed.
6. The Integration with Azure DevOps is not something we use yet. We are entering a situation where its going to be used. Even then not always.....for end-to-end tests there is no 1-to-1 correlation between test results and a specific branch. 
7. We need LivingDoc 's ability to filter on outcomes.
8.  The analysis on step-definitions for us adds no value as they are shared over multiple projects. Its normal for a step-definition to not be used in a test run.
10. We use the `ISpecFlowOutputHelper` to include  OpenTelemetry TraceId's and links to logfiles (Blobstorage / S3 etc etc) into the LivingDoc report.

**Reply by gasparnagy** [reply]
Thx. The "filter on outcomes" is especially a good point that I did not think of. Also good to know your feedback about the analysis on step-definitions. I had the same feeling.

**Comment by Digbigpig** [comment]
Hi I'd like to chime in
In the past i've had a great experience using specflow and specrunner+. I'm now setting up a new test automation solution and realized we are without a reporting feature in reqnroll. Like others have mentioned, it could be a reason to delay reqnroll adoption over specflow.

The main feature we need is the ability to generate an html report that can be easily navigated and parsed. We'd like it to support navigation/ filtering between test categories. We'd like for easy screenshot integration into the report. Being able to see the step, the timing, and console output for each test is an expected feature.

I come from using spec runner and that was everything I needed. The livingdoc stuff looks cool and has nice features but some of it seems overkill for my general use case, however I never used it besides failing to integrate it with reqnroll :)

PS reqnroll sounds closer to rick-roll than rock n roll to me🕺

**Reply by gasparnagy** [reply]
There is no realistic chance that we will be able to redo the SpecFlow+ Runner. What would be good though is to make the livingdoc generator HTML "template" more extensible, so one could make a more "classic test report" kind of template for it (we can even put it to the product itself).

**Comment by brajpal-jangir** [comment]
Thanks for bringing this in discussion!

We used CLI command to generate reports. These are integrated in DevOps pipelines.
-	In the active development phase, we generate living doc without results to let Product owner and other stakeholders review the test coverage. Living doc also have links to UserStory (from TFS) to validate the requirement.
-	In times of GA, we generate living doc with results and archive them as test metrics as part of GA.

I also tried to use SpecFlow+LivingDoc DevOps extension, but it has some limitations and that’s where I switched to use CLI. Few limitations which I came across are
-	We have multiple Specflow Projects in GitRepo and it’s not possible to publish multiple Living doc for single git branch. It will always override the previous uploaded document for same git branch.
-	We also came across the issue when Living document HTML file size is more and it failed to publish document on DevOps. I reported this issue but am not able to find the link now.

**Comment by MathiasLopes** [comment]
I've found a way to generate LivingDoc HTML from Reqnroll unit tests. I've made a GitHub project to explain it, here's the link if you're interested: 
https://github.com/MathiasLopes/ReqnrollTemporaryLivingDoc

**Reply by Digbigpig** [reply]
Interesting Mathias! I'll take a look later and let you know what I think

**Reply by ShanmukhaGajula** [reply]
@MathiasLopes , I have cloned your repo and tried to execute but looks like it is generating the TestExecution.json but however livingdoc is not picking up the test results.

**Reply by MathiasLopes** [reply]
@ShanmukhaGajula You have to execute the following command line : 
livingdoc feature-folder . --output-type HTML -t bin/Debug/net8.0/TestExecution.json
from the folder where csproj is located
Follow the readme and you should get the .html file
If you want, you can add in post-build event these lines : 
dotnet test --no-build
livingdoc feature-folder . --output-type HTML -t bin/Debug/net8.0/TestExecution.json
To launch automatically unit tests and the build of the living doc after each build. 
However, this may slow down builds

**Reply by ShanmukhaGajula** [reply]
@MathiasLopes  I followed the exact steps from the your readme.md, I am getting .html file generated but it is not showing me the execution results and yes I have executed this command but no details of the execution
livingdoc feature-folder . --output-type HTML -t bin/Debug/net8.0/TestExecution.json

**Reply by ShanmukhaGajula** [reply]
@MathiasLopes , small correction here. I gave the command as livingdoc feature-folder "mypath to features folder" --output-type HTML -t bin/Debug/net8.0/TestExecution.json instead of livingdoc feature-folder . --output-type HTML -t bin/Debug/net8.0/TestExecution.json but after removing the path it worked. Thanks again for the workaround. Appreciate your inputs.

**Reply by dominik-niebuhr** [reply]
> @MathiasLopes , small correction here. I gave the command as livingdoc feature-folder "mypath to features folder" --output-type HTML -t bin/Debug/net8.0/TestExecution.json instead of livingdoc feature-folder . --output-type HTML -t bin/Debug/net8.0/TestExecution.json but after removing the path it worked. Thanks again for the workaround. Appreciate your inputs.

Have you got the test results to show up?

**Reply by bartziengs** [reply]
Thank you very much for sharing this. I can confirm that this works for our use case and that I have successfully generated the living doc documentation in existing DevOps pipelines for reqnroll projects by using the [Build Step]( https://docs.specflow.org/projects/specflow-livingdoc/en/latest/Generating/Configuring-the-Build-Step-in-YAML.html ) and configuring generator source to 'FeatureFolder' (which under the hood runs the command you described).

**Reply by RajeshPoola** [reply]
> Thank you very much for sharing this. I can confirm that this works for our use case and that I have successfully generated the living doc documentation in existing DevOps pipelines for reqnroll projects by using the [Build Step](https://docs.specflow.org/projects/specflow-livingdoc/en/latest/Generating/Configuring-the-Build-Step-in-YAML.html) and configuring generator source to 'FeatureFolder' (which under the hood runs the command you described).

I'm having difficulty getting the results published through the Azure pipeline. I used the task below, which is converted into the command shown when the pipeline is executed. It seems to be adding the default path automatically. Could you please advise me if I'm missing anything?

I am getting warning messages and results are not published
**_Warning: No feature files were discovered._**

  - task: techtalk.techtalk-specflow-plus.specflow-plus.SpecFlowPlus@0
    displayName: Generate SpecFlow living documentation
    inputs:
      generatorSource: 'FeatureFolder'
      projectFilePath: .
      testExecutionJson: '**/*TestExecution*.json'
    condition: succeededOrFailed()

dotnet "D:\a\_tasks\SpecFlowPlus_32f3fe66-8bfc-476e-8e2c-9b4b59432ffa\0.6.964\CLI\LivingDoc.CLI.dll" feature-folder "D:\a\1\s" --output-type JSON --test-execution-json "**/*TestExecution*.json"  --include-test-output "All" --output "D:\a\1\s\19238\FeatureData.json" --project-language "en"

**Reply by Insire** [reply]
> I've found a way to generate LivingDoc HTML from Reqnroll unit tests. I've made a GitHub project to explain it, here's the link if you're interested: https://github.com/MathiasLopes/ReqnrollTemporaryLivingDoc

Thanks for that sample. I can confirm that it works well for NUnit. I also successfully converted the sample so that it uses xunit instead. However i am unable to get it to work in an internal project with xunit, when all my tests pass. The `Testexecution.json` isnt being generated for me, when i have no tests failing.

I see the sample project references `databinding.LivingDoc.ExecutionDataCollector` can you clarify where the code for that nuget package is or whether its even available to the public? It does not seem to be part of the official reqnroll github repository and there is no link on the package page on nuget.org. I'm interested to take a look at it and figure out how it works and what i may be doing incorrectly.

Alternatively if you have some advice on how to enable logging or similar diagnostics, that would also be very much appreciated.

**Reply by tula-engineered-quality-hub** [reply]
Hi @Insire , I did another project for proof of concept. The report was generated successfully, both with tests passing and failing in the same execution, and with all tests passing (in this case, I ran it with 1 because for testing I removed the other one that I implemented to fail and do my proof of concept). Here is a print of the csproj, test executed and livingdo c command used to generate the documentation: 

![image](https://github.com/user-attachments/assets/2a9bc758-8393-4164-b13f-78daae5855a5)

But I'm having trouble publishing this via pipe release so that I can display the generated HTML in Azure, it could even be on a wiki page. What I need to solve is that stakeholders have access to the documentation in a centralized way in Azure, like the livingdoc extension did (before it was discontinued and refused authentication). One alternative I thought of is to generate the report via the Livingdoc CLI command line in the release pipe yml after running the tests, save it in a folder based on the repository and branch in S3, and somehow retrieve this HTML via the S3 URL and display this report on a wiki page, I don't know if by pointing to a given URL or via "embed" to take advantage of the wiki's own widget that enables this display of embed URLs, at least temporarily until we have an extension for Reqnroll that can do this. Anyway, has anyone had this problem and solved it, even if temporarily in some way? If so, do you have a tutorial or step-by-step guide?

**Reply by Insire** [reply]
> Hi @Insire , I did another project for proof of concept. The report was generated successfully, both with tests passing and failing in the same execution, and with all tests passing (in this case, I ran it with 1 because for testing I removed the other one that I implemented to fail and do my proof of concept). Here is a print of the csproj, test executed and livingdo c command used to generate the documentation:
> 
> ![image](https://private-user-images.githubusercontent.com/108248997/400953620-2a9bc758-8393-4164-b13f-78daae5855a5.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzYzMjI1NDksIm5iZiI6MTczNjMyMjI0OSwicGF0aCI6Ii8xMDgyNDg5OTcvNDAwOTUzNjIwLTJhOWJjNzU4LTgzOTMtNDE2NC1iMTNmLTc4ZGFhZTU4NTVhNS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwMTA4JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDEwOFQwNzQ0MDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kMGIxNmQxYzllNDVhZmYzNDk2ZmE2OTBmZWJjZGE5Mzg0NDFmMmJkMDIwY2NhYWI1YTNjOWNmMWU1YmRhZDE4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.m_F6fYNWcovwxLG7xyVlbkqRPF2w48xG1R_iR4Mv53Y)
> 
> But I'm having trouble publishing this via pipe release so that I can display the generated HTML in Azure, it could even be on a wiki page. What I need to solve is that stakeholders have access to the documentation in a centralized way in Azure, like the livingdoc extension did (before it was discontinued and refused authentication). One alternative I thought of is to generate the report via the Livingdoc CLI command line in the release pipe yml after running the tests, save it in a folder based on the repository and branch in S3, and somehow retrieve this HTML via the S3 URL and display this report on a wiki page, I don't know if by pointing to a given URL or via "embed" to take advantage of the wiki's own widget that enables this display of embed URLs, at least temporarily until we have an extension for Reqnroll that can do this. Anyway, has anyone had this problem and solved it, even if temporarily in some way? If so, do you have a tutorial or step-by-step guide?

We are using the [azure devops specflow extension](https://marketplace.visualstudio.com/items?itemName=techtalk.techtalk-specflow-plus) which seems to pick up the livingdoc html file, when i upload it has artifact during CI/CD build. I am already manually creating the html file via the [specflow cli](https://www.nuget.org/packages/SpecFlow.Plus.LivingDoc.CLI) and upload it to the current build via [cake](https://cakebuild.net/), which uses [this](https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#upload-upload-an-artifact) under the hood.

**Reply by tula-engineered-quality-hub** [reply]
And is it possible to authenticate with the Microsoft account from the livingdoc extension? I was able to do it until the end of December, but not anymore: 

![image](https://github.com/user-attachments/assets/4dde9829-f34b-4144-97ed-dceb930f9433)
![image](https://github.com/user-attachments/assets/de4663ac-6de7-40b7-9a1f-87f1508de5ed)

**Reply by RajeshPoola** [reply]
They pulled the plug on Specflow https://support-hub.tricentis.com/open?id=post&number=NEW0001432

**Reply by tula-engineered-quality-hub** [reply]
Yes @RajeshPoola. That's why I came to change this idea of ​​what you're doing to view the post-deploy report.

**Reply by RajeshPoola** [reply]
@tulavalle They are currently not allowing new signups or logins. Mine is still working as I signed up a long time ago. However, if my browser or Azure forces me to reauthenticate, it could become a problem for me too.
![image](https://github.com/user-attachments/assets/a107edf5-a21c-416d-be44-c1cad51650bd)

**Reply by rdvanbuuren-gasunie** [reply]
Did anyone manage to run livingdoc with the `test-assembly` option and .NET 9?
I'm getting the following error:
```bash
Tool 'specflow.plus.livingdoc.cli' (version '3.9.57') was successfully installed.
Framework: .NET 8.0.12
Warning: Unable to load one or more of the requested types.
Could not load file or assembly 'System.Runtime, Version=9.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'. The system cannot find the file specified.
```
I noticed that `Framework: .NET 8.0.12` is being selected, however the project is .NET 9. I can reproduce this locally so I'm wondering if I can force the Framework somehow. According to the [NuGet](https://www.nuget.org/packages/SpecFlow.Plus.LivingDoc.CLI#supportedframeworks-body-tab) it should also support .NET 9 and in one pipeline run I've seen it work. Maybe someone knows how it picks the framework?

**Reply by gasparnagy** [reply]
I haven't tried, but you can try with setting the `DOTNET_ROLL_FORWARD` env variable to `LatestMajor` (or maybe `Major`) and run it again.

**Comment by ShanmukhaGajula** [comment]
Yes I do.. thanks for the input. One question how can i add base64 image to
the reporter?

**Comment by Pressacco** [comment]
# Status Update

As I understand it, the current game-plan is:
- to be compatible with the _Cucumber_ eco-system by supporting [Cucumber messages](https://github.com/cucumber/messages)
   - in order to integrate with tools like [Pickles](https://www.picklesdoc.com/?utm_source=chatgpt.com), [Report Portal](https://github.com/reportportal/reportportal), etc.
- which is being tracked by #233

yes?

**Reply by clrudolphi** [reply]
Yes. Support for Cucumber Messages is the first step.

**Comment by gasparnagy** [comment]
FYI: https://reqnroll.net/news/2025/08/reqnroll-v3-0-released-with-reporting/

**Reply by CRohm42** [reply]
Thank you for your hard work!  This is awesome!

**Reply by jrod567** [reply]
Way to go!!  Looks great - thank you so much!

**Reply by SierraNL** [reply]
Very cool, but we're running into a CSP issue, causing empty pages. Because the generated pages use inline scripts. Is it possible to add a hash/nonce to the pages? Or should I create an issue for that?

**Reply by clrudolphi** [reply]
> Very cool, but we're running into a CSP issue, causing empty pages. Because the generated pages use inline scripts. Is it possible to add a hash/nonce to the pages? Or should I create an issue for that?

Yes , please submit an issue.

**Comment by blaisep** [comment]
Gaspar, I'm sorry for the delayed response. I have very little practical
experience with ndjson, and for HTML output I typically use Python (or
pandoc)
Python handles ndjson beautifully, but I'm not here to sell you on Python.

My dear friend Lisa Crispin thinks highly of you, and I am a fan of
cucumber; so I am eager to contribute in any way I can.

You will be hearing more from me.

Blaise

LinkedIn <https://www.linkedin.com/in/blaisepabon/>   |  Github
<https://github.com/blaisep>
“If you want to go fast, go alone. If you want to go far, go
together.” --African
proverb

**Comment by AndreasLauvhjell** [comment]
Good to see version 3.0 already here with living doc. However, my .html file is empty. It gets created but it's 0 KB. Is .NET8 required or something? We run on .NET Framework 4.8

**Reply by clrudolphi** [reply]
Please submit a bug report on this.

**Reply by AndreasLauvhjell** [reply]
I fixed it by installing Microsoft.NET.Test.Sdk. Works flawlessly now.

**Comment by nithinbandaru1** [comment]
Feedback:

Report:

1. Scenario Outline - I don't see the multiple results. I see only one. Not sure if it is limitation of Cucumber messages itself or the integration.
   - `Update` - I see it works, just that it's not easy to realize being a new user 😊. I have to hover and see the mouse pointer changing.
2. Server, Git, Tags as UNKNOWN at top of document. I did not see them in sample at https://reports.cucumber.io/.
3. No partial search for the tags.
4. No duration's for the steps, background etc.
5. No count of pass, fail, skip at overall level and just percentage is shown.
6. Good to have pass, fail, skip percentage and count at each level.
7. Good to have dynamic hierarchy customization instead of flattening to features. With this I could list scenarios or group by folders by levels. Nothing new I see this in code coverage reports.

Usage:

1. I see we can use reqnroll.json or environment variable. I am just wondering about the possibility of having separate NuGets (One for HTML, One for cucumber messages) and on addition just auto register and produce report.
2. I see that path always works only relative to bin folder. How can I provide the path in such a way that it will be relative to execution location. Probably if we specify ./ then consider relative to current path where being executed, if /x/ or x/ consider relative to bin folder.

**Reply by clrudolphi** [reply]
> Feedback:

Thanks @nithinbandaru1 for taking the time to review and provide feedback.

>Server, Git, Tags as UNKNOWN at top of document. I did not see them in sample at https://reports.cucumber.io/.

These depend upon our ability to detect which build server is being used. If you are running these locally on your workstation without a build server, then these UNKNOWN values are expected. The reports.cucumber.io server is using a customized version of the renderer and they seem to have omitted the build info. We do not yet have built-in integration with the cucumber reports server.

>No partial search for the tags.
>No duration's for the steps, background etc.
>No count of pass, fail, skip at overall level and just percentage is shown.
>Good to have pass, fail, skip percentage and count at each level.
>Good to have dynamic hierarchy customization instead of flattening to features. With this I could list scenarios or group by folders by levels. 

These would all be behaviors of the Cucumber HtmlFormatter and React rendering components (which we consume from Cucumber). We'll take these feedback items and pass them along to the Cucumber team for consideration of enhancements.

>I see we can use reqnroll.json or environment variable. I am just wondering about the possibility of having separate NuGets

We considered this early in the development and decided instead to simply build these two basic formatters into the core project. Other formatters can be custom built and added as plugins, but the configuration of them would be similar (reqnroll.json or environment variables).

>I see that path always works only relative to bin folder. 

Can you provide an example? The default directory location for the output is the currently executing assembly directory (eg, `<Project>/bin/<Configuration>/<framework>`, such as `myproject/bin/Debug/net9.0`)

**Comment by hung-doan** [comment]
Thank you for the great work on v3.0 and the new reporting mechanism!
I’d like to ask if there are any plans to support a "step definition report".

In SpecFlow's LivingDoc, it’s possible to generate a StepDefinitionReport.html, which is extremely helpful for providing the team with a clear overview of all available steps.

The current report mainly focuses on test execution results, which is valuable, but not as useful for this particular use case.

**Reply by gasparnagy** [reply]
This is not in our short term plans, but if you choose the message formatter, the generated cucumber messages ndjson file contains all the info for that, so you (or someone) can easily make a tool that extracts that.

**Comment by mohamedilyeshel** [comment]
Hello, Thank you for releasing the report feature to reqnroll.
I have a question, is it possible to integrate the generated report (html version or .ndjson) into azure devops ?
because the extension Specflow + livingDoc is no longer available.
Is there a way to do this?

**Reply by gasparnagy** [reply]
I succeeded to do that by generating the HTML report and using this plugin to show it: https://marketplace.visualstudio.com/items?itemName=awardedsolutions.azure-pipelines-html-report-awardedsolutions

I will try to find time to make some description here about the steps, but it was not hard to get that working.

**Reply by ghost** [reply]
I would like to do something similar. Can you please provide some details on how to get this working? Thanks

**Comment by kwlin** [comment]
**tl;dr;**
Does anybody have experience with another tools (licensed is oke) for generating Living docs files based on cucumber messages for large code bases & applications? The tool/site should really focus for functional users/readers.



Hi,

We are currently researching a way to migrate our feature & test result files from the old Pickles tool (https://github.com/picklesdoc/pickles) to cucumber messages, cucumber-htmlformatter & cucumber-reactcomponents. 

Our current workflow with Pickles is 
* aggregate all the feature & test result files from **various** testruns and run the Pickles tool
* We push the html output to a our living doc site (static content site)

The html output is very user/reader friendly because it has grouping on features (with treeview & folders) & has a search functionality on all the content.

When using cucumber messages, cucumber-htmlformatter & cucumber-reactcomponents:
**Pros:**
* Default support "& standardized way of working(?)"
* Search works like a charm (on all content)
* It's a great start for small applications or low feature sets.
**Cons:**
* The reqnroll - htmlformatter only generates per cucumber message a html page; There is a workaround: Writing our own code to merge all the cucumber messages & generate the html page.
* The html page from cucumber-reactcomponents is a simple list of all the scenario's. So when we have a lot of features & test results, the html page will be very large and hard to find a specific group of features. 


Alternative we have also research [ExpressiumOSS](https://github.com/ExpressiumOSS/Expressium.LivingDoc):
**Pros:**
* Features are nicely grouped 
* Possible to merge various feature & testresult files
**Cons**
* Search only works on feature title & tags
* 3rd party tool/library with still 1 active developer

Does anybody have experience with other tooling for living documentations with focus on readability & user friendly?

**Reply by egorshulga** [reply]
At our project we adopted Allure reports

**Reply by hung-doan** [reply]
using Allure reports as well. So far, they’ve been ok, and the Reqnroll integration has been well documented

**Reply by MarcWils** [reply]
We use feature descriptions to describe additional context and business rules. Am I missing something? I can't find this in the Allure reports. 

<img width="1237" height="683" alt="image" src="https://github.com/user-attachments/assets/0d5ae3ec-3242-4ef6-ae92-3aa506136260" />

**Reply by priveLupus** [reply]
We startet using the HTML Report, but it just doesn't feel the same as the Specflow+ LivingDoc.
Also we miss the easy access via the Overview in Azure. Navigating via the Builds seems scary to non-technical people in our team.
Does Allure offer something like that?
